gax/src/main/java/com/google/api/gax/batching/FlowControlEventStats.java

```diff
@@ -0,0 +1,133 @@
+ * the future. The events can be used to dynamically adjust concurrency in the client. For example:
+ *
+ * <pre>{@code
+ * // Increase flow control limits if there was throttling in the past 5 minutes and throttled time
+ * // was longer than 1 minute.
+ * while(true) {
+ *    FlowControlEvent event = flowControlEventStats.getLastFlowControlEvent();
+ *    if (event != null
+ *         && event.getTimestampMs() > System.currentMillis() - TimeUnit.MINUTES.toMillis(5)
+ *         && event.getThrottledTimeInMs() > TimeUnit.MINUTES.toMillis(1)) {
+ *      flowController.increaseThresholds(elementSteps, byteSteps);
+ *    }
+ *    Thread.sleep(TimeUnit.MINUTE.toMillis(10));
+ * }
+ * }</pre>
+ */
+@InternalApi("For google-cloud-java client use only")
+public class FlowControlEventStats {
+
+  private volatile FlowControlEvent lastFlowControlEvent;
+
+  // We only need the last event to check if there was throttling in the past X minutes, so this
+  // doesn't need to be super accurate.
+  void recordFlowControlEvent(FlowControlEvent event) {
+    if (lastFlowControlEvent == null || event.compareTo(lastFlowControlEvent) > 0) {
+      lastFlowControlEvent = event;
+    }
+  }
+
+  public FlowControlEvent getLastFlowControlEvent() {
+    return lastFlowControlEvent;
+  }
+
+  /**
+   * A flow control event. Record throttled time if {@link LimitExceededBehavior} is {@link
+   * LimitExceededBehavior#Block}, or the exception if the behavior is {@link
+   * LimitExceededBehavior#ThrowException}.
+   */
+  @AutoValue
+  public abstract static class FlowControlEvent implements Comparable<FlowControlEvent> {
+    static FlowControlEvent createReserveDelayed(long throttledTimeInMs) {
+      return createReserveDelayed(System.currentTimeMillis(), throttledTimeInMs);
+    }
+
+    static FlowControlEvent createReserveDenied(FlowControlException exception) {
+      return createReserveDenied(System.currentTimeMillis(), exception);
+    }
+
+    /** Package-private for use in testing. */
+    @VisibleForTesting
+    static FlowControlEvent createReserveDelayed(long timestampMs, long throttledTimeInMs) {
+      Preconditions.checkArgument(timestampMs > 0, "timestamp must be greater than 0");
+      Preconditions.checkArgument(throttledTimeInMs > 0, "throttled time must be greater than 0");
+      return new AutoValue_FlowControlEventStats_FlowControlEvent(
+          timestampMs, throttledTimeInMs, null);
+    }
+
+    /** Package-private for use in testing. */
+    @VisibleForTesting
+    static FlowControlEvent createReserveDenied(long timestampMs, FlowControlException exception) {
+      Preconditions.checkArgument(timestampMs > 0, "timestamp must be greater than 0");
+      Preconditions.checkArgument(
+          exception != null, "FlowControlException can't be null when reserve is denied");
+      return new AutoValue_FlowControlEventStats_FlowControlEvent(timestampMs, null, exception);
+    }
+
+    public abstract long getTimestampMs();
```

Comment:
Subjective, but it looks like this class does not need to be `AutoValue`. It is way less trivial than most of the `AutoValue` classes, which are usually just a set of setters/getters (not the case here). Please consider just storing the `throttledtimeInMs`, `exception` and `timestampMs` values manually in this class.

Reply:
agreed

Reply:
AutoValue doesnt actually hurt anything here. I generally prefer AutoValue here to communicate the fact that this is a simple value type